transport/happyeyeballs.go

```diff
@@ -0,0 +1,237 @@
+func (d *HappyEyeballsStreamDialer) DialStream(ctx context.Context, addr string) (StreamConn, error) {
+	host, port, err := net.SplitHostPort(addr)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse address: %w", err)
+	}
+	if net.ParseIP(host) != nil {
+		// Host is already an IP address, just dial the address.
+		return d.dial(ctx, addr)
+	}
+
+	// Indicates to attempts that the search is done, so they don't get stuck.
+	searchCtx, searchDone := context.WithCancel(ctx)
+	defer searchDone()
+
+	// DOMAIN NAME LOOKUP SECTION
+	// We start the IPv4 and IPv6 lookups in parallel, writing to lookup4Ch
+	// and lookup6Ch when they are done.
+	type LookupResult struct {
+		IPs []netip.Addr
+		Err error
+	}
+	lookup6Ch := make(chan LookupResult)
+	lookup4Ch := make(chan LookupResult)
+	go func(lookup6Ch chan<- LookupResult, host string) {
+		ips, err := d.lookupIPv6(searchCtx, host)
+		if err != nil {
+			err = fmt.Errorf("failed to lookup IPv6 addresses: %w", err)
+		}
+		lookup6Ch <- LookupResult{ips, err}
+		close(lookup6Ch)
+	}(lookup6Ch, host)
+	go func(lookup4Ch chan<- LookupResult, host string) {
+		ips, err := d.lookupIPv4(searchCtx, host)
+		if err != nil {
+			err = fmt.Errorf("failed to lookup IPv4 addresses: %w", err)
+		}
+		lookup4Ch <- LookupResult{ips, err}
+		close(lookup4Ch)
+	}(lookup4Ch, host)
+
+	// DIAL ATTEMPTS SECTION
+	// We keep IPv4s and IPv6 separate and track the last one attempted so we can
+	// alternate the address family in the connection attempts.
+	var ip4s []netip.Addr
+	var ip6s []netip.Addr
+	var lastDialed netip.Addr
+	// Keep track of the lookup and dial errors separately. We prefer the dial errors
+	// when returning.
+	var lookupErr error
+	var dialErr error
+	// Channel to wait for before a new dial attempt. It starts
+	// with a closed channel that doesn't block because there's no
+	// wait initially.
+	var attemptDelayCh <-chan struct{} = newClosedChan()
+	type DialResult struct {
+		Conn StreamConn
+		Err  error
+	}
+	var dialCh = make(chan DialResult)
+
+	// Channel that triggers when a new connection can be made. Starts blocked (nil)
+	// because we need IPs first.
+	var readyToDialCh <-chan struct{} = nil
+	// We keep track of pending operations (lookups and IPs to dial) so we can stop when
+	// there's no more work to wait for.
+	for opsPending := 2; opsPending > 0; {
+		if len(ip6s) == 0 && len(ip4s) == 0 {
+			// No IPs. Keep dial disabled.
+			readyToDialCh = nil
+		} else {
+			// There are IPs to dial.
+			if !lastDialed.IsValid() && len(ip6s) == 0 && lookup6Ch != nil {
+				// Attempts haven't started and IPv6 lookup is not done yet. Set up Resolution Delay, as per
+				// https://datatracker.ietf.org/doc/html/rfc8305#section-8, if it hasn't been set up yet.
+				if readyToDialCh == nil {
+					resolutionDelayCtx, cancelResolutionDelay := context.WithTimeout(searchCtx, 50*time.Millisecond)
+					defer cancelResolutionDelay()
+					readyToDialCh = resolutionDelayCtx.Done()
```

Comment:
I'm assuming you meant `time.NewTimer(50*time.Milisecond).C`. I wasn't sure if we needed to drain the `Timer.C` channel, but it seems it's buffered and we don't need to?

In any case, I tried based on your suggestion, but the type of channel is different (`<-chan time.Time`), so it requires more changes, not really simpler. Also, the timer sends the time once, versus a context closes the channel, which is easier to reason about (a read will never block).

Another important point is that I still need to stop the timer, so I need to pass it to the goroutine to call Stop in case the dial returns before the timer triggers.

On the race condition, that's a valid point. A cancellation will potentially trigger an extra dial. I tried different ways to trigger, but couldn't. In any case, I changed the code to use the background context instead.